lib/irb/type_completion/completor.rb

```diff
@@ -0,0 +1,238 @@
+          if scope.local_variables.include?(name)
+            value_doc.call scope[name]
+          else
+            method_doc.call scope.self_type
+          end
+        else
+        end
+      end
+
+      def candidates_from_result(result)
+        candidates = case result
+        in [:require | :require_relative => method, name]
+          if IRB.const_defined? :RegexpCompletor # IRB::VERSION >= 1.8.2
+            path_completor = IRB::RegexpCompletor.new
+          elsif IRB.const_defined? :InputCompletor # IRB::VERSION <= 1.8.1
+            path_completor = IRB::InputCompletor
+          end
+          if !path_completor
+            []
+          elsif method == :require
+            path_completor.retrieve_files_to_require_from_load_path
+          else
+            path_completor.retrieve_files_to_require_relative_from_current_dir
+          end
+        in [:call_or_const, type, name, self_call]
+          ((self_call ? type.all_methods : type.methods).map(&:to_s) - HIDDEN_METHODS) | type.constants
+        in [:const, type, name, scope]
+          if type
+            scope_constants = type.types.flat_map do |t|
+              scope.table_module_constants(t.module_or_class) if t.is_a?(Types::SingletonType)
+            end
+            (scope_constants.compact | type.constants.map(&:to_s)).sort
+          else
+            scope.constants.sort | ReservedWords
+          end
+        in [:ivar, name, scope]
+          ivars = scope.instance_variables.sort
+          name == '@' ? ivars + scope.class_variables.sort : ivars
+        in [:cvar, name, scope]
+          scope.class_variables
+        in [:gvar, name, scope]
+          scope.global_variables
+        in [:symbol, name]
+          Symbol.all_symbols.map { _1.inspect[1..] }
+        in [:call, type, name, self_call]
+          (self_call ? type.all_methods : type.methods).map(&:to_s) - HIDDEN_METHODS
+        in [:lvar_or_method, name, scope]
+          scope.self_type.all_methods.map(&:to_s) | scope.local_variables | ReservedWords
+        else
+          []
+        end
+        [name || '', candidates]
+      end
+
+      def analyze(code, binding = Object::TOPLEVEL_BINDING)
+        # Workaround for https://github.com/ruby/prism/issues/1592
+        return if code.match?(/%[qQ]\z/)
+
+        lvars_code = binding.local_variables.map do |name|
+          "#{name}="
+        end.join + "nil;\n"
+        code = lvars_code + code
+        ast = Prism.parse(code).value
+        name = code[/(@@|@|\$)?\w*[!?=]?\z/]
+        *parents, target_node = find_target ast, code.bytesize - name.bytesize
+        return unless target_node
+
+        calculate_scope = -> { TypeAnalyzer.calculate_target_type_scope(binding, parents, target_node).last }
+        calculate_type_scope = ->(node) { TypeAnalyzer.calculate_target_type_scope binding, [*parents, target_node], node }
+
+        if target_node.is_a?(Prism::StringNode) || target_node.is_a?(Prism::InterpolatedStringNode)
```

Comment:
We should be able to utilise `Prism::Dispatcher` and avoid these `is_a?(NodeClass)` conditions and the case statement below. You can use this [test](https://github.com/ruby/prism/blob/ac693764ae14c293f800611ce4e23e60dfcc64d7/test/prism/dispatcher_test.rb#L6) as an example.

Reply:
Combined this `is_a?` into `case target_node when`.